mathics/core/pattern.py

```diff
@@ -1199,3 +1218,51 @@ def per_name(yield_name: Callable, groups: Tuple, vars_dict: dict):
     # def yield_name(setting):
     #    yield_func(setting)
     per_name(yield_choice, tuple(groups.items()), vars_dict)
+
+
+# FIXME: return type should be a specific kind of Tuple, not a tuple.
+def pattern_sort_key(pat) -> tuple:
+    """
+    This function builds the generic sort key used when
+    patterns and rules must be sorted in the evaluation order, i.e.
+    more specific patterns should come before the most general.
+
+    Keys are sorted following the logic of Python tuples order:
+    comparisons are done element by element until the i-th element
+    of the first tuple is smaller than the i-th element of the 
+    second one. If the elements are equal, comparison continues
+    until it finishes with the shortest tuple. In that case, the shortest
+    tuple is considered the smaller one.
+
+    Pattern sort key structure:
+        0: 0/2:        Atom / Expression
+        1: pattern:    0 / 11-31 for blanks / 1 for empty Alternatives /
+                           40 for OptionsPattern
+        2: 0/1:        0 for PatternTest
+        3: 0/1:        0 for Pattern
+        4: 0/1:        1 for Optional
+        5: head / 0 for atoms
+        6: elements / 0 for atoms
+        7: 0/1:        0 for Condition
+    """
+    if isinstance(pat, AtomPattern):
+        return ATOM_PATTERN_SORT_KEY
+
+    # Append (4,) to elements so that longer expressions have higher
+    # precedence
+    result = (
+        2,
+        0,
+        1,
+        1,
+        0,
+        pat.head.get_sort_key(True),
+        tuple(
```

Comment:
TODO: to reproduce the WMA evaluation order, we need to add an element between the head and the elements that indicates if the pattern matches with a definite number of elements (leaves) or if it has `BlankSequence`-like elements. The goal is that `Expr[___]` comes after `Expr[_,_]`.